src/providers/postgres/qgscolumntypethread.cpp

```diff
@@ -67,43 +67,48 @@ void QgsGeomColumnTypeThread::run()
     return;
   }
 
-  int i = 0, n = layerProperties.size();
-  for ( QVector<QgsPostgresLayerProperty>::iterator it = layerProperties.begin(),
-        end = layerProperties.end();
-        it != end; ++it )
+  int totalLayers = layerProperties.length();
+  int addedLayers = 0;
+
+  emit progress( addedLayers, totalLayers );
+
+  QList<QgsPostgresLayerProperty> unrestrictedLayers;
+
+  if ( ! dontResolveType )
   {
-    QgsPostgresLayerProperty &layerProperty = *it;
-    if ( !mStopped )
+    for ( auto layerProperty : layerProperties )
     {
-      emit progress( i++, n );
-      emit progressMessage( tr( "Scanning column %1.%2.%3…" )
-                            .arg( layerProperty.schemaName,
-                                  layerProperty.tableName,
-                                  layerProperty.geometryColName ) );
-
       if ( !layerProperty.geometryColName.isNull() &&
            ( layerProperty.types.value( 0, QgsWkbTypes::Unknown ) == QgsWkbTypes::Unknown ||
              layerProperty.srids.value( 0, std::numeric_limits<int>::min() ) == std::numeric_limits<int>::min() ) )
       {
-        if ( dontResolveType )
-        {
-          QgsDebugMsg( QStringLiteral( "skipping column %1.%2 without type constraint" ).arg( layerProperty.schemaName, layerProperty.tableName ) );
-          continue;
-        }
-
-        mConn->retrieveLayerTypes( layerProperty, mUseEstimatedMetadata );
+        unrestrictedLayers << layerProperty;
+      }
+      else
+      {
+        // Immediately tell the layer list dialog box about
+        // the completed layers
+        emit setLayerType( layerProperty );
+        emit progress( ++addedLayers, totalLayers );
       }
     }
+  }
 
-    if ( mStopped )
-    {
-      layerProperty.types.clear();
-      layerProperty.srids.clear();
-      break;
-    }
+  if ( unrestrictedLayers.empty() || mStopped )
+  {
+    emit progress( 0, 0 );
+    emit progressMessage( mStopped ? tr( "Table retrieval stopped." ) : tr( "Table retrieval finished." ) );
+    QgsPostgresConnPool::instance()->releaseConnection( mConn );
+    mConn = nullptr;
+  }
 
-    // Now tell the layer list dialog box...
+  mConn->retrieveLayerTypes( unrestrictedLayers, mUseEstimatedMetadata );
+
+  for ( auto layerProperty : unrestrictedLayers )
```

Comment:
```suggestion
  for ( const auto &layerProperty : qgis::as_const(unrestrictedLayers) )
```

untested, but you get the idea, please also do the same for other iterators in this PR.

Reply:
Done, as part of an interactive rebase to master branch, to reorganize commits in logical steps.
I've also found the reason why tables where not shown, hopefully next travis build will be green (your test was great!)